### PR TITLE
doc: Update mpirun map/rank/bind to reflect PRRTE

### PR DESCRIPTION
**Pull request by jjhursey**

### CODE REVIEW

**Comment by jjhursey** [comment]
This is still a WIP, but I'm trying to write down the defaults with examples for PRRTE (per conversation [here](https://github.com/open-mpi/ompi/issues/10734#issuecomment-1295414593) ). While working on the examples we found some bugs in PRRTE that have been fixed. This documentation reflects the behavior of PRRTE [head](https://github.com/openpmix/prrte/commit/a61d6a88bebfc9e32d9b8ee9ce1f9ec51bb20cb3) working inside Open MPI `main`.

**Comment by jsquyres** [comment]
❤️

**Comment by jjhursey** [comment]
I'm done with my pass through the `mpirun` man page. I focused on map/bind/rank but checked a few other options. The examples work and are accurate with the current PRRTE `master` which includes the following fixes I found along the way:
 * https://github.com/openpmix/prrte/pull/1567
 * https://github.com/openpmix/prrte/pull/1568
 * https://github.com/openpmix/prrte/pull/1569
 * https://github.com/openpmix/prrte/pull/1577
 * https://github.com/openpmix/prrte/pull/1578
 * https://github.com/openpmix/prrte/pull/1579
 * https://github.com/openpmix/prrte/pull/1583

**Comment by jjhursey** [comment]
Ref: https://github.com/open-mpi/ompi/wiki/Webex-affinity-discussions-2019-09

File: docs/man-openmpi/man1/mpirun.1.rst [hunk 1]
```diff
@@ -1052,40 +1015,140 @@ less convenient than that of ``mpirun`` options.  On the other hand,
...
+
+.. code::
+
+  shell$ mpirun --np 4 --host hostA:4,hostB:2 ./a.out
+  R0  hostA  [BB/BB/BB/BB/BB/BB/BB/BB][../../../../../../../..]
+  R1  hostA  [BB/BB/BB/BB/BB/BB/BB/BB][../../../../../../../..]
+  R2  hostA  [../../../../../../../..][BB/BB/BB/BB/BB/BB/BB/BB]
+  R3  hostA  [../../../../../../../..][BB/BB/BB/BB/BB/BB/BB/BB]
+
+If only ``--map-by OBJ`` is specified then it implies ``--bind-to OBJ --rank-by fill``. The example below results in ``--map-by hwthread --bind-to hwthread --rank-by fill``
```
Comment (jsquyres) [inline docs/man-openmpi/man1/mpirun.1.rst:1119]:
```suggestion
If only ``--map-by OBJ`` is specified, then it implies ``--bind-to OBJ --rank-by fill``. The example below results in ``--map-by hwthread --bind-to hwthread --rank-by fill``
```

File: docs/man-openmpi/man1/mpirun.1.rst [hunk 2]
```diff
@@ -1052,40 +1015,140 @@ less convenient than that of ``mpirun`` options.  On the other hand,
...
+
+.. code::
+
+  shell$ mpirun --np 4 --map-by hwthread --host hostA:4,hostB:2 ./a.out
+  R0  hostA  [B./../../../../../../..][../../../../../../../..]
+  R1  hostA  [.B/../../../../../../..][../../../../../../../..]
+  R0  hostA  [../B./../../../../../..][../../../../../../../..]
+  R1  hostA  [../.B/../../../../../..][../../../../../../../..]
+
+If only ``--bind-to OBJ`` is specified then ``--map-by`` is determined by the number of processes and ``--rank-by fill``. The example below results in ``--map-by package --bind-to core --rank-by fill``
```
Comment (jsquyres) [inline docs/man-openmpi/man1/mpirun.1.rst:1129]:
```suggestion
If only ``--bind-to OBJ`` is specified, then ``--map-by`` is determined by the number of processes and ``--rank-by fill``. The example below results in ``--map-by package --bind-to core --rank-by fill``
```

File: docs/man-openmpi/man1/mpirun.1.rst [hunk 3]
```diff
@@ -1269,12 +1332,12 @@ Process Environment
...
 
 Processes in the MPI application inherit their environment from the
-Open RTE daemon upon the node on which they are running.  The
+PRRTE daemon upon the node on which they are running.  The
 environment is typically inherited from the user's shell.  On remote
 nodes, the exact environment is determined by the boot MCA module
 used.  The rsh launch module, for example, uses either rsh/ssh to
-launch the Open RTE daemon on remote nodes, and typically executes one
-or more of the user's shell-setup files before launching the Open RTE
+launch the PRTE daemon on remote nodes, and typically executes one
```
Comment (jsquyres) [inline docs/man-openmpi/man1/mpirun.1.rst:1339]:
```suggestion
launch the PRRTE daemon on remote nodes, and typically executes one
```

File: docs/man-openmpi/man1/mpirun.1.rst [hunk 4]
```diff
@@ -1269,12 +1332,12 @@ Process Environment
...
+or more of the user's shell-setup files before launching the PRTE
```
Comment (jsquyres) [inline docs/man-openmpi/man1/mpirun.1.rst:1340]:
```suggestion
or more of the user's shell-setup files before launching the PRRTE
```

**Comment by rhc54** [comment]
@jjhursey Would you be willing to port the example sections over to the PRRTE help files (section [placement-examples] in help-prte-rmaps-base.txt) as well? If you don't have time, but wouldn't object to someone doing it, would it be okay if I do so?

File: docs/man-openmpi/man1/mpirun.1.rst [hunk 5]
```diff
@@ -286,59 +279,66 @@ To map processes:
 
 * ``--map-by <object>``: Map to the specified object, defaults to
   socket. Supported options include ``slot``, ``hwthread``, ``core``,
-  ``L1cache``, ``L2cache``, ``L3cache``, ``socket``, ``numa``,
-  ``board``, ``node``, ``sequential``, ``distance``, and ``ppr``. Any
-  object can include modifiers by adding a ``:`` and any combination
-  of ``PE=n`` (bind n processing elements to each proc), ``SPAN``
+  ``L1cache``, ``L2cache``, ``L3cache``, ``package``, ``numa``,
+  ``node``, ``seq``, ``rankfile``, ``pe-list=#``, and ``ppr``. 
+  Note that ``socket`` is also accepted as an alias for ``package``.
```
[the remark this answers is not in the document]
Comment (jjhursey) [reply docs/man-openmpi/man1/mpirun.1.rst:284]: Good suggestion. Will add this in the next push.

File: docs/man-openmpi/man1/mpirun.1.rst [hunk 6]
```diff
@@ -286,59 +279,66 @@ To map processes:
...
+  Any object can include modifiers by adding a ``:`` and any combination
+  of ``pe=n`` (bind n processing elements to each proc), ``span``
```
[the remark this answers is not in the document]
Comment (jjhursey) [reply docs/man-openmpi/man1/mpirun.1.rst:286]: Will do

**Comment by jjhursey** [comment]
@jsquyres I just pushed an update with your suggestions.

File: docs/man-openmpi/man1/mpirun.1.rst [hunk 7]
```diff
@@ -286,59 +279,68 @@ To map processes:
...
+
+    * ``pe=n`` : bind n processing elements to each proc
+    * ``span`` : load balance the processes across the allocation
+    * ``oversubscribe`` : allow more processes on a node than processing elements
+    * ``nooversubscribe`` : (default)
+    * ``nolocal`` : do not place processes on the same host as the `mpirun` process
+    * ``hwtcpus`` : use hardware threads as CPU slots for mapping
+    * ``corecpus`` : use processor cores as CPU slots for mapping (Default)
+    * ``file=`` : used with ``rankfile`` to specify the file to use
+    * ``ordered`` : used with ``pe-list`` to bind each process to one of the specifed processing elements
```
Comment (jsquyres) [inline docs/man-openmpi/man1/mpirun.1.rst:295]: Nit picky suggestions:

```suggestion
    * ``pe=n``: bind ``n`` processing elements to each proc
    * ``span``: load balance the processes across the allocation
    * ``oversubscribe``: allow more processes on a node than processing elements
    * ``nooversubscribe``: do *not* allow more processes on a node than processing elements (default)
    * ``nolocal``: do not place processes on the same host as the ``mpirun`` process
    * ``hwtcpus``: use hardware threads as CPU slots for mapping
    * ``corecpus``: use processor cores as CPU slots for mapping (default)
    * ``file=filename``: used with ``rankfile``; use ``filename`` to specify the file to use
    * ``ordered``: used with ``pe-list`` to bind each process to one of the specified processing elements
```

File: docs/man-openmpi/man1/mpirun.1.rst [hunk 8]
```diff
@@ -1052,40 +1029,140 @@ less convenient than that of ``mpirun`` options.  On the other hand,
...
+    * Result: ``--map-by OBJ --bind-to OBJ --rank-by fill``
+
+* If only ``--bind-to OBJ`` (where ``OBJ`` is something like ``core``) is specified, then:
+
+    * ``--map-by`` is either ``core`` or ``package`` depending on the number of processes
+    * ``--bind-to`` specified ``OBJ``
+    * ``--rank-by`` defaults to ``fill``
+    * Result: ``--map-by OBJ --bind-to OBJ --rank-by fill``
+
+* If ``--map-by OBJ1 --bind-to OBJ2``, then:
```
Comment (jsquyres) [inline docs/man-openmpi/man1/mpirun.1.rst:1106]: RST requires a blank line after this line (before the sub-bullets).

**Comment by jjhursey** [comment]
Good catch. Fixed.

**Review by jsquyres — APPROVED** [review]

**Pull request merged**